File: src/edu/stanford/cfuller/analysistoolsinterface/SegmentationController.java
/* ***** BEGIN LICENSE BLOCK *****
 * 
 * Copyright (c) 2011 Colin J. Fuller
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * 
 * ***** END LICENSE BLOCK ***** */


package edu.stanford.cfuller.analysistoolsinterface;

import edu.stanford.cfuller.imageanalysistools.frontend.AnalysisController;
import edu.stanford.cfuller.imageanalysistools.frontend.DataSummary;
import edu.stanford.cfuller.imageanalysistools.method.Method;
import edu.stanford.cfuller.imageanalysistools.parameters.ParameterDictionary;
import edu.stanford.cfuller.imageanalysistools.parameters.ParameterParserFactory;
import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.prefs.Preferences;
import javax.swing.DefaultComboBoxModel;
import javax.swing.JFileChooser;
import javax.swing.filechooser.FileNameExtensionFilter;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

/**
 *
 * @author cfuller
 */
public class SegmentationController extends TaskController implements OmeroListener{

    final static String METHOD_XML_FILENAME = "edu/stanford/cfuller/analysistoolsinterface/resources/methods.xml";
    final static String METHOD_XML_TAG = "method";
    final static String DISPLAY_ATTR_NAME = "displayname";
    final static String CLASS_ATTR_NAME= "class";


    final static String STATUS_PROCESSING = "Processing";
    final static String STATUS_READY = "Ready";
    final static String STATUS_SUMMARY = "Making data summary";
	final static String STATUS_OMERO_ERR = "OMERO missing; ";
    
    SegmentationWindow sw;

    List<Long> omeroImageIds;
    OmeroBrowsingWindowController omeroBrowser;


    @Override
    public void startTask() {

        this.omeroBrowser = null;

        sw = new SegmentationWindow(this);
        sw.addWindowListener(this);
        initializeMethods();

        String imageFilename = Preferences.userNodeForPackage(this.getClass()).get("imageFile", "");
        String parameterFilename = Preferences.userNodeForPackage(this.getClass()).get("parameterFile", "");
        int selectedMethodIndex = Preferences.userNodeForPackage(this.getClass()).getInt("selectedIndex", 0);

        sw.setImageFilename(imageFilename);
        sw.setParameterFilename(parameterFilename);
        sw.setSelectedMethodIndex(selectedMethodIndex);

        LoggingUtilities.addHandler(sw.getLogHandler());

        this.sw.setStatus(STATUS_READY);

        sw.setVisible(true);



    }

    @Override
    public void imageIdsHaveBeenSelected(List<Long> ids) {
        this.omeroImageIds = ids;
        this.sw.setUseOmeroServer(true);
    }

    public void browseForParameterFile() {
        String path = this.sw.getParameterFilename();
        JFileChooser fc = null;
        if ((new File(path)).exists()) {
            fc = new JFileChooser(path);
        } else {
            fc = new JFileChooser();
        }
        fc.setFileSelectionMode(JFileChooser.FILES_ONLY);
        fc.addChoosableFileFilter(new FileNameExtensionFilter("XML Files", "xml"));
		fc.addChoosableFileFilter(new FileNameExtensionFilter("ruby scripts", "rb"));
        int retVal = fc.showOpenDialog(this.sw);


        if (retVal == JFileChooser.APPROVE_OPTION) {

            String selected = fc.getSelectedFile().getAbsolutePath();


            sw.setParameterFilename(selected);
        }
    }

    public void browseForImageFileOrDirectory() {
        String path = this.sw.getImageFilename();
        JFileChooser fc = null;
        if ((new File(path)).exists()) {
            fc = new JFileChooser(path);
        } else {
            fc = new JFileChooser();
        }
        fc.setFileSelectionMode(JFileChooser.FILES_AND_DIRECTORIES);
        
        int retVal = fc.showOpenDialog(this.sw);

        if (retVal == JFileChooser.APPROVE_OPTION) {

            String selected = fc.getSelectedFile().getAbsolutePath();


            sw.setImageFilename(selected);
        }
    }


    public void onMethodSelect() {
        int index = this.sw.getSelectedMethodIndex();

        MethodInfo mi = (MethodInfo) this.sw.getMethodComboBoxModel().getElementAt(index);

        if (mi.getMethodClass() == null) {
            this.sw.setCustomMethodFieldEnabled(true);
        } else {
            this.sw.setCustomMethodFieldEnabled(false);
        }

    }

    public void useOmeroDataSource() {
		try {
        	Class.forName("omero.api.GatewayPrx");
		} catch (ClassNotFoundException e) {
			LoggingUtilities.warning("Could not open OMERO data source; OMERO client plugin missing.");
			this.sw.setStatus(STATUS_OMERO_ERR + STATUS_READY);
			this.sw.disableOmero();
			return;
		}
		this.omeroBrowser = new OmeroBrowsingWindowController();
    	omeroBrowser.openNewBrowsingWindow(this);
    }

    public void runButtonPressed() {

        if (! (this.sw.getStatus().equals(STATUS_READY) || this.sw.getStatus().equals(STATUS_OMERO_ERR + STATUS_READY))) return;


        final String parameterFilename = this.sw.getParameterFilename();
        final String imageFilename = this.sw.getImageFilename();

        Preferences.userNodeForPackage(this.getClass()).put("imageFile", imageFilename);
        Preferences.userNodeForPackage(this.getClass()).put("parameterFile", parameterFilename);
        Preferences.userNodeForPackage(this.getClass()).putInt("selectedIndex", this.sw.getSelectedMethodIndex());

		this.sw.setStatus(STATUS_PROCESSING);

		java.awt.EventQueue.invokeLater(new Runnable() {
			
			public void run() {
				
				MethodInfo mi = (MethodInfo) sw.getMethodComboBoxModel().getElementAt(sw.getSelectedMethodIndex());

		        Class c = mi.getMethodClass();
				
				final ParameterDictionary pd = ParameterParserFactory.createParameterParserForFile(parameterFilename).parseFileToParameterDictionary(parameterFilename);        

		        if (c == null && !pd.hasKey("method_name")) {
		            try {
		                c = Class.forName(sw.getCustomClassName());
		            } catch (ClassNotFoundException e) {
		                LoggingUtilities.warning("Custom class not found with name: " + sw.getCustomClassName());
						sw.setStatus(STATUS_READY);
		                return;
		            } 
		        }

		        pd.addIfNotSet("method_name", c.getName());
		        if ((new File(imageFilename)).isDirectory()) {
		            pd.addIfNotSet("local_directory", imageFilename);
		        } else {
		            File wholeFilename = new File(imageFilename);
		            String name = wholeFilename.getName();
		            String dir = wholeFilename.getParent();
		            if (!pd.hasKey("common_filename_tag")) {pd.addIfNotSet("common_filename_tag", name);
		            } else {
		                pd.setValueForKey("image_extension", name);
		            }
		            pd.addIfNotSet("local_directory", dir);
		        }

		        pd.addIfNotSet("max_threads", Integer.toString(Runtime.getRuntime().availableProcessors()));

		        if (sw.getUseOmeroServer()) {
		            pd.setValueForKey("use_omero", "true");
		            pd.addIfNotSet("omero_hostname", omeroBrowser.getHostname());
		            pd.addIfNotSet("omero_username", omeroBrowser.getUsername());
		            pd.addIfNotSet("omero_password", omeroBrowser.getPassword());
		            String omeroImageIdString = null;
		            for (long l : omeroImageIds) {
		                if (omeroImageIdString == null) {
		                    omeroImageIdString = "";
		                } else {
		                    omeroImageIdString += " ";
		                }
		                omeroImageIdString += Long.toString(l);
		            }

		            pd.addIfNotSet("omero_image_ids", omeroImageIdString);
		        }


		       	if (! sw.summarizeDataOnly()) {
                   AnalysisController ac = new AnalysisController();
                   ac.addAnalysisLoggingHandler(sw.getLogHandler());
                   ac.runLocal(pd);
				}
				
				sw.setStatus(STATUS_SUMMARY);
				
				try {
				    DataSummary.SummarizeData(pd.getValueForKey("local_directory")+File.separator+AnalysisController.DATA_OUTPUT_DIR, pd.getValueForKey("local_directory")+File.separator+AnalysisController.PARAMETER_OUTPUT_DIR);
				} catch (IOException e) {
				    LoggingUtilities.severe("Encountered error while summarizing data.");
				}
				
				sw.setStatus(STATUS_READY);
	
			}
		});
                        
        

        

    }




    protected static class MethodInfo {
        String displayName;
        Class methodClass;

        public MethodInfo(String displayName, String className) {
            this.displayName = displayName;
            if (className != null) {
                try {
                    this.methodClass =  Class.forName(className);
                } catch (ClassNotFoundException e) {
                    LoggingUtilities.warning("Valid class not found for method with name: " + displayName);
                    this.methodClass = null;
                } 
            } else {
                this.methodClass = null;
            }
        }

        public String getDisplayName() {
            return this.displayName;
        }

        public Class getMethodClass() {
            return this.methodClass;
        }

        @Override
        public String toString() {
            return this.displayName;
        }

    }

    protected void initializeMethods() {
	
        DefaultComboBoxModel model = sw.getMethodComboBoxModel();

        String methodResourceLocation = this.getClass().getClassLoader().getResource(METHOD_XML_FILENAME).toString();

        NodeList methodNodes = null;

        try {
            methodNodes = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(methodResourceLocation).getElementsByTagName(METHOD_XML_TAG);
        } catch (SAXException e) {
            LoggingUtilities.severe("Encountered exception while parsing tasks xml file.");
            return;
        } catch (ParserConfigurationException e) {
            LoggingUtilities.severe("Incorrectly configured xml parser.");
            return;
        } catch (IOException e) {
            LoggingUtilities.severe("Exception while reading xml file.");
            return;
        }

        for (int i =0; i < methodNodes.getLength(); i++) {
            Node n = methodNodes.item(i);

            Node displayName = n.getAttributes().getNamedItem(DISPLAY_ATTR_NAME);
            Node className = n.getAttributes().getNamedItem(CLASS_ATTR_NAME);

            model.addElement(new MethodInfo(displayName != null? displayName.getNodeValue() : null, className != null? className.getNodeValue() : null));
        }


    }




}

File: src/main/java/edu/stanford/cfuller/analysistoolsinterface/SegmentationController.kt
package edu.stanford.cfuller.analysistoolsinterface

import edu.stanford.cfuller.imageanalysistools.frontend.AnalysisController
import edu.stanford.cfuller.imageanalysistools.frontend.DataSummary
import edu.stanford.cfuller.imageanalysistools.method.Method
import edu.stanford.cfuller.imageanalysistools.meta.parameters.ParameterDictionary
import edu.stanford.cfuller.imageanalysistools.meta.AnalysisMetadata
import edu.stanford.cfuller.imageanalysistools.meta.AnalysisMetadataParserFactory
import java.io.File
import java.io.IOException
import java.util.prefs.Preferences
import javax.swing.DefaultComboBoxModel
import javax.swing.JFileChooser
import javax.swing.filechooser.FileNameExtensionFilter
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import org.xml.sax.SAXException

/**

 * @author cfuller
 */
class SegmentationController : TaskController(), OmeroListener {
    internal var sw: SegmentationWindow = SegmentationWindow(this)
    internal var omeroImageIds: List<Long> = listOf()
    internal var omeroBrowser: OmeroBrowsingWindowController? = null

    override fun startTask() {
        this.omeroBrowser = null
        sw = SegmentationWindow(this)
        sw.addWindowListener(this)
        initializeMethods()
        val imageFilename = Preferences.userNodeForPackage(this::class.java).get("imageFile", "")
        val parameterFilename = Preferences.userNodeForPackage(this::class.java).get("parameterFile", "")
        val selectedMethodIndex = Preferences.userNodeForPackage(this::class.java).getInt("selectedIndex", 0)
        sw.imageFilename = imageFilename
        sw.parameterFilename = parameterFilename
        sw.selectedMethodIndex = selectedMethodIndex
        LoggingUtilities.addHandler(sw.logHandler)
        this.sw.status = STATUS_READY
        sw.isVisible = true
    }

    override fun imageIdsHaveBeenSelected(ids: List<Long>) {
        this.omeroImageIds = ids
        this.sw.useOmeroServer = true
    }

    fun browseForParameterFile() {
        val path = this.sw.parameterFilename
        var fc: JFileChooser? = null
        if (File(path).exists()) {
            fc = JFileChooser(path)
        } else {
            fc = JFileChooser()
        }
        fc.fileSelectionMode = JFileChooser.FILES_ONLY
        fc.addChoosableFileFilter(FileNameExtensionFilter("Parameter files (.xml, .rb)", "xml", "rb"))
        val retVal = fc.showOpenDialog(this.sw)


        if (retVal == JFileChooser.APPROVE_OPTION) {

            val selected = fc.selectedFile.absolutePath


            sw.parameterFilename = selected
        }
    }

    fun browseForImageFileOrDirectory() {
        val path = this.sw.imageFilename
        var fc: JFileChooser? = null
        if (File(path).exists()) {
            fc = JFileChooser(path)
        } else {
            fc = JFileChooser()
        }
        fc.fileSelectionMode = JFileChooser.FILES_AND_DIRECTORIES

        val retVal = fc.showOpenDialog(this.sw)

        if (retVal == JFileChooser.APPROVE_OPTION) {

            val selected = fc.selectedFile.absolutePath


            sw.imageFilename = selected
        }
    }


    fun onMethodSelect() {
        val index = this.sw.selectedMethodIndex

        val mi = this.sw.methodComboBoxModel.getElementAt(index) as MethodInfo

        if (mi.methodClass == null) {
            this.sw.setCustomMethodFieldEnabled(true)
        } else {
            this.sw.setCustomMethodFieldEnabled(false)
        }

    }

    fun useOmeroDataSource() {
        try {
            Class.forName("omero.api.GatewayPrx")
        } catch (e: ClassNotFoundException) {
            LoggingUtilities.warning("Could not open OMERO data source; OMERO client plugin missing.")
            this.sw.status = STATUS_OMERO_ERR + STATUS_READY
            this.sw.disableOmero()
            return
        }

        this.omeroBrowser = OmeroBrowsingWindowController()
        omeroBrowser!!.openNewBrowsingWindow(this)
    }

    fun runButtonPressed() {

        if (!(this.sw.status == STATUS_READY || this.sw.status == STATUS_OMERO_ERR + STATUS_READY)) return


        val parameterFilename = this.sw.parameterFilename
        val imageFilename = this.sw.imageFilename

        Preferences.userNodeForPackage(this.javaClass).put("imageFile", imageFilename)
        Preferences.userNodeForPackage(this.javaClass).put("parameterFile", parameterFilename)
        Preferences.userNodeForPackage(this.javaClass).putInt("selectedIndex", this.sw.selectedMethodIndex)

        this.sw.status = STATUS_PROCESSING

        java.awt.EventQueue.invokeLater(Runnable {
            val mi = sw.methodComboBoxModel.getElementAt(sw.selectedMethodIndex) as MethodInfo

            var c: Class<*>? = mi.methodClass

            val am = AnalysisMetadataParserFactory.createParserForFile(parameterFilename).parseFileToAnalysisMetadata(parameterFilename)

            val pd = am.inputParameters

            if (c == null && !pd!!.hasKey("method_name")) {
                try {
                    c = Class.forName(sw.customClassName)
                } catch (e: ClassNotFoundException) {
                    LoggingUtilities.warning("Custom class not found with name: " + sw.customClassName)
                    sw.status = STATUS_READY
                    return@Runnable
                }

            }

            pd!!.addIfNotSet("method_name", c!!.name)
            if (File(imageFilename).isDirectory) {
                pd.addIfNotSet("local_directory", imageFilename)
            } else {
                val wholeFilename = File(imageFilename)
                val name = wholeFilename.name
                val dir = wholeFilename.parent
                if (!pd.hasKey("common_filename_tag")) {
                    pd.addIfNotSet("common_filename_tag", name)
                } else {
                    pd.setValueForKey("image_extension", name)
                }
                pd.addIfNotSet("local_directory", dir)
            }

            pd.addIfNotSet("max_threads", Integer.toString(Runtime.getRuntime().availableProcessors()))

            if (sw.useOmeroServer) {
                pd.setValueForKey("use_omero", "true")
                pd.addIfNotSet("omero_hostname", omeroBrowser!!.hostname)
                pd.addIfNotSet("omero_username", omeroBrowser!!.username)
                pd.addIfNotSet("omero_password", omeroBrowser!!.getPassword())
                var omeroImageIdString: String? = null
                for (l in omeroImageIds) {
                    if (omeroImageIdString == null) {
                        omeroImageIdString = ""
                    } else {
                        omeroImageIdString += " "
                    }
                    omeroImageIdString += java.lang.Long.toString(l)
                }

                pd.addIfNotSet("omero_image_ids", omeroImageIdString!!)
            }


            Thread(Runnable {
                if (!sw.summarizeDataOnly()) {
                    val ac = AnalysisController()
                    ac.addAnalysisLoggingHandler(sw.logHandler)
                    ac.runLocal(am)
                }

                sw.status = STATUS_SUMMARY

                try {
                    DataSummary.SummarizeData(pd.getValueForKey("local_directory") + File.separator + AnalysisController.DATA_OUTPUT_DIR, pd.getValueForKey("local_directory") + File.separator + AnalysisController.PARAMETER_OUTPUT_DIR)
                } catch (e: IOException) {
                    LoggingUtilities.severe("Encountered error while summarizing data.")
                }

                sw.status = STATUS_READY
            }).start()
        })
    }

    private class MethodInfo(displayName: String, className: String?) {
        var displayName: String
            internal set
        var methodClass: Class<*>? = null
            internal set

        init {
            this.displayName = displayName
            if (className != null) {
                try {
                    this.methodClass = Class.forName(className)
                } catch (e: ClassNotFoundException) {
                    LoggingUtilities.warning("Valid class not found for method with name: " + displayName)
                    this.methodClass = null
                }

            } else {
                this.methodClass = null
            }
        }

        override fun toString(): String {
            return this.displayName
        }

    }

    private fun initializeMethods() {
        val model = sw.methodComboBoxModel
        val methodResourceLocation = this.javaClass.classLoader.getResource(METHOD_XML_FILENAME)!!.toString()
        var methodNodes: NodeList? = null

        try {
            methodNodes = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(methodResourceLocation).getElementsByTagName(METHOD_XML_TAG)
        } catch (e: SAXException) {
            LoggingUtilities.severe("Encountered exception while parsing tasks xml file.")
            return
        } catch (e: ParserConfigurationException) {
            LoggingUtilities.severe("Incorrectly configured xml parser.")
            return
        } catch (e: IOException) {
            LoggingUtilities.severe("Exception while reading xml file.")
            return
        }

        for (i in 0..methodNodes!!.length - 1) {
            val n = methodNodes.item(i)
            val displayName = n.attributes.getNamedItem(DISPLAY_ATTR_NAME)
            val className = n.attributes.getNamedItem(CLASS_ATTR_NAME)
            model.addElement(MethodInfo(displayName?.nodeValue!!, className?.nodeValue).toString())
        }
    }

    companion object {
        internal val METHOD_XML_FILENAME = "edu/stanford/cfuller/analysistoolsinterface/resources/methods.xml"
        internal val METHOD_XML_TAG = "method"
        internal val DISPLAY_ATTR_NAME = "displayname"
        internal val CLASS_ATTR_NAME = "class"
        internal val STATUS_PROCESSING = "Processing"
        internal val STATUS_READY = "Ready"
        internal val STATUS_SUMMARY = "Making data summary"
        internal val STATUS_OMERO_ERR = "OMERO missing; "
    }
}
